File: 1/2021.03.11/36.Valid Sudoku/java/solution.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        List<Map<Character, Integer>> row = new ArrayList<>();
        List<Map<Character, Integer>> col = new ArrayList<>();
        List<Map<Character, Integer>> box = new ArrayList<>();
        for (int i = 0; i < 9; i++) {
            row.add(new HashMap<>());
            col.add(new HashMap<>());
            box.add(new HashMap<>());
        }
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != '.') {
                    row.get(i).put(board[i][j], row.get(i).getOrDefault(board[i][j], 0) + 1);
                    if (row.get(i).get(board[i][j]) > 1) {
                        return false;
                    }
                    col.get(j).put(board[i][j], col.get(j).getOrDefault(board[i][j], 0) + 1);
                    if (col.get(j).get(board[i][j]) > 1) {
                        return false;
                    }
                    int boxIndex = (j / 3) * 3 + i / 3;
                    box.get(boxIndex).put(board[i][j], box.get(boxIndex).getOrDefault(board[i][j], 0) + 1);
                    if (box.get(boxIndex).get(board[i][j]) > 1) {
                        return false;
                    }
                }
            }
        }
        return true;
    }
}

File: 1/2021.03.11/36.Valid Sudoku/python/solution.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row = [{} for _ in range(9)]
        col = [{} for _ in range(9)]
        box = [{} for _ in range(9)]
        for i in range(9):
            for j in range(9):
                if '1' <= board[i][j] <= '9':
                    row[i][board[i][j]] = row[i].get(board[i][j], 0) + 1
                    if row[i][board[i][j]] > 1:
                        return False
                    col[j][board[i][j]] = col[j].get(board[i][j], 0) + 1
                    if col[j][board[i][j]] > 1:
                        return False
                    box_index = (j // 3) * 3 + i // 3
                    box[box_index][board[i][j]] = box[box_index].get(board[i][j], 0) + 1
                    if box[box_index][board[i][j]] > 1:
                        return False 
        return True
